File: host-libs/lua/loaders/js_deno.lua
--- JS Deno (V8) loader registration for polyplug.
local ffi = require("ffi")

local ok = pcall(ffi.cdef, [[
    typedef struct { uint8_t _reserved; } PolyplugJsDenoConfig;
    void* polyplug_js_deno_loader_create(const PolyplugJsDenoConfig* cfg);
    void  polyplug_js_deno_loader_free(void* ptr);
    uint32_t polyplug_runtime_register_loader(void* rt, void* loader);
]])

local _lib = nil
local function get_lib()
    if not _lib then
        _lib = ffi.load("polyplug_js_deno")
    end
    return _lib
end

local M = {}

function M.register(rt)
    local lib = get_lib()
    local cfg = ffi.new("PolyplugJsDenoConfig", {0})
    local loader = lib.polyplug_js_deno_loader_create(cfg)
    if loader == nil then
        error("polyplug: js_deno loader create failed")
    end
    local polyplug_lib = ffi.load("polyplug")
    local err = polyplug_lib.polyplug_runtime_register_loader(rt, loader)
    if err ~= 0 then
        error("polyplug: js_deno loader register failed: " .. tostring(err))
    end
end

return M

File: host-libs/js/loaders/js_deno.ts
let _lib: Deno.DynamicLibrary<typeof JS_DENO_SYMBOLS> | null = null;

const JS_DENO_SYMBOLS = {
    polyplug_js_deno_loader_create: {
        parameters: ["pointer"] as const,
        result: "pointer" as const,
    },
} satisfies Deno.ForeignLibraryInterface;

function getLib() {
    if (!_lib) {
        const libPath = Deno.env.get("POLYPLUG_JS_DENO_LIB")
            ?? "libpolyplug_js_deno.so";
        _lib = Deno.dlopen(libPath, JS_DENO_SYMBOLS);
    }
    return _lib;
}

export function registerJsDenoLoader(
    rt: Deno.PointerValue,
    registerFn: (rt: Deno.PointerValue, loader: Deno.PointerValue) => number
): void {
    const lib = getLib();
    const cfgBuf = new Uint8Array([0]);
    const cfgPtr = Deno.UnsafePointer.of(cfgBuf);
    const loaderPtr = lib.symbols.polyplug_js_deno_loader_create(cfgPtr);
    if (loaderPtr === null) {
        throw new Error("polyplug: js_deno loader create failed");
    }
    const err = registerFn(rt, loaderPtr);
    if (err !== 0) {
        throw new Error(`polyplug: js_deno loader register failed: ${err}`);
    }
}
